pkgs/tools/misc/colorless/default.nix

```diff
@@ -0,0 +1,47 @@
+{ lib
+, fetchurl
+, stdenvNoCC
+, coreutils
+, dash
+, gnused
+, less
+}:
+
+stdenvNoCC.mkDerivation rec {
+  pname = "colorless";
+  version = "109";
+
+  src = fetchurl {
+    url = "http://software.kimmo.suominen.com/${pname}-${version}.tar.gz";
+    sha256 = "039a140fa11cf153cc4d03e4f753b7ff142cab88ff116b7600ccf9edee81927c";
+  };
+
+  makeFlags = [
+    "TOOLPATH=${coreutils}/bin:${gnused}/bin:${less}/bin"
+    "PREFIX=${placeholder "out"}"
+    "SHELL=${dash}/bin/dash"
```

Comment:
Does it required dash to build?

Reply:
It requires a POSIX shell to **run**. My preference for a POSIX shell is `dash` for its smaller size and its ready availability (e.g. on NetBSD there is no `bash` in the base).

On non-Nixpkgs I would just use `#!/bin/sh` and I was not sure at all about what the proper approach would be with Nixpkgs. For example, [on pkgsrc I used this](http://cvsweb.netbsd.org/bsdweb.cgi/pkgsrc/misc/colorless/Makefile?rev=1.3&content-type=text/x-cvsweb-markup&only_with_tag=MAIN):

```
MAKE_FLAGS+=		SHELL=${TOOLS_PLATFORM.sh}
```

For the Debian packaging I just get `SHELL=/bin/sh` from `make` (although I will be setting it explicitly in the next version of `colorless`).

Reply:
On any system using nix, it's essentially guaranteed that bash is already in the nix store. Also, `/bin/sh` is bash on nixos systems. So using dash is adding to the installed size in practice on most systems. Maybe a `.override`able option to allow the user's choice of shell, with a default of bash, would be a decent solution?

Reply:
For the size I was thinking more about RAM than storage.

But using the "default `/bin/sh`" of the platform is a sound default, and I would like that.

@tejing1: Could you show me how to add the option?  I'm very new to Nixpkgs.

Reply:
```diff
diff --git a/pkgs/tools/misc/colorless/default.nix b/pkgs/tools/misc/colorless/default.nix
index f8c37b4d157..97a34219cf9 100644
--- a/pkgs/tools/misc/colorless/default.nix
+++ b/pkgs/tools/misc/colorless/default.nix
@@ -2,7 +2,8 @@
 , fetchurl
 , stdenvNoCC
 , coreutils
-, dash
+, bash
+, preferredPosixShell ? "${bash}/bin/bash"
 , gnused
 , less
 }:
@@ -19,7 +20,7 @@ stdenvNoCC.mkDerivation rec {
   makeFlags = [
     "TOOLPATH=${lib.makeBinPath [ coreutils gnused less ]}"
     "PREFIX=$(out)"
-    "SHELL=${dash}/bin/dash"
+    "SHELL=${preferredPosixShell}"
   ];
 
   preInstall = ''
```
Or so. I'm not 100% sure if there's a more globally accepted name for the option.

Reply:
Thank you. I used `binSh` after some `grep`ping.  I also saw `environment.binsh` and it made me wonder if I could or should have somehow tied into that.

Reply:
You can't tie to a nixos option in a package, I'm afraid. Other than by referencing `/bin/sh` directly, anyway, but depending on well-known paths is discouraged when packaging for nix. `/bin/sh` only exists on nixos systems as a compromise.

If there was a global way to configure this, it would be through nixpkgs config, but there doesn't seem to be an accepted option for it there. This kind of "use flag-like" config seems to have fallen out of favor, and was never well standardized, anyway.